[CORE-9764] pandaproxy/sr: Revert to an eager evaluation model on sr startup

IoannisRP opened the pull request:
Implements: [CORE-9764](https://redpandadata.atlassian.net/browse/CORE-9764)

Behavioral changes:
- A invalid schema stored in the _schemas topic will still be visible to the user and could be retrieved by id or subject/version lookups.
- When submitting a normalized schema, the stored schemas will not be re-normalized.

Notes:
There is a regression in behavior compared to the lazy loading. Schemas that have forward references (they have a reference to a schema further down the topic that hasn't been loaded yet) will be treated as invalid. This is being addressed in #25768, which builds on this.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x
- [ ] v24.1.x

## Release Notes

* none


[CORE-9764]: https://redpandadata.atlassian.net/browse/CORE-9764?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
pandaproxy/sr: Merge unparsed and canonical schema types to common type
called subject schema

IoannisRP committed:
pandaproxy/sr: Rename subject schema type to stored schema

IoannisRP committed:
pandaproxy/sr: Re-introduce an eager schema loading

Re-introduce an attempt to make schema canonical during the
initialization phase of the schema registry. Unlike before,
if the canonicalization fails we will store the schema as-is in
the topic.

IoannisRP committed:
pandaproxy/sr: Rollback to previous behavior

This commit is mostly a rollback of #f6f43cf23894a6a134beeb069e5144cd4920d009

Changes:
- Kept new tests added in that commit.
- Remove redundant validation call. The schema passed in is already
validated on the caller side.

IoannisRP committed:
pandaproxy/sr: Move back to eager evaluation on sr initialization

This is a functional rollback of PR #25217.
However there are some key changes.

Changes:
- We keep all the new tests as they document the behavior of edge cases.
- Unlike before, schemas in the topic are always loaded in the store,
  even if they are invalid (missing references, syntax errors, etc...).
- Schema lookup by id or subject/version will always return the stored
  schema, regardless if it is valid or invalid. Note that this doesn't
  affect lookup by schema, which behaves like before. An invalid input
  schema will fail to be processed, it will throw an error and the
  request will fail.

rockwotj commented on src/transform-sdk/cpp/examples/schema_registry.cc:
You should not be needing to make any changes in `src/transform-sdk`

BenPope commented on src/v/pandaproxy/api/api-doc/schema_registry.json:
```diff
@@ -1203,4 +1203,4 @@
     }
-  
\ No newline at end of file
+  
```
This newline is unwanted

IoannisRP replied:
This was not a real extra new line. It was an annoying `<eof>` finishing character and what vim thought was appropriate for `<eof>`. Sorting this problem took more time than the tests and the rest combined :sweat_smile:

BenPope commented on src/v/pandaproxy/schema_registry/types.h:
```diff
@@ -190,35 +186,14 @@ class typed_schema_definition {
     references _refs;
 };
 
-///\brief An unvalidated definition of the schema and its type.
-///
-/// This comes from the user and should be considered as potentially
-/// ill-formed.
-using unparsed_schema_definition
-  = typed_schema_definition<struct unparsed_schema_defnition_tag>;
-
-///\brief A canonical definition of the schema and its type.
-///
-/// This form is stored on the topic and returned to the user.
-using canonical_schema_definition
-  = typed_schema_definition<struct canonical_schema_definition_tag>;
```
Can you explain why this distinction is no longer useful? It was put in place to prevent accidentally storing unvalidated input.

If anything, it might be nice to have new types for `normalized` or `serialized` (in the future).

An alternative would be to express the form with a value that takes part in the equality check, but in general I prefer to push this type of logical error to compile time.

IoannisRP replied:
I wouldn't say it is not useful. I would say it causes more trouble than it's worth.

Since the lazy evaluation (taken away here and brought back in #25768), we don't only store validated schemas in the store. We store valid and invalid schemas next to each other. To do this with this type distinction you need to either use `unparsed`, which is not true for valid stored schemas, or you need to use a variant or some other equivalent method.

However, in every case, we treat valid and invalid schemas the same. For the places that need to know if a schema is valid or not (like compatibility checks), this is done on-the-fly by validating the schema. The only place that it matters to sanitize our inputs is 2 endpoints that accept a schema as an argument.

I don't think that 2 functions that need input sanitization is worth it to bring complication to the whole codebase.

If we wanted a type to ensure that input sanitization has happened, i would argue that we need a type with very limited effect (used only in the `sharded_store` api that performs schema lookups). The rest of the SR just cares about a "schema". It doesn't care about it's validity.

IoannisRP replied:
I don't agree we need a distinction between `serialized` and `normalized` schemas, as they are in every sense equal (when the rest of the equality holds).

Specifically, if you post a schema in it's normalized form with `normalize=false`, thus a `sanitized` schema, you still want it to match an equivalent schema that you posted previously with `normalize=true`.

BenPope commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -2774,6 +2781,13 @@ def test_subjects_subject_versions(entry,
                                            expected_successful=True)
 
+        #forward_references will fail the lookup, as they didn't get canonicallized.
+        #Each will create a new entree.
```
Now I'm hungry.
```suggestion
        #Each will create a new entry.
```

BenPope commented on src/v/pandaproxy/schema_registry/sharded_store.cc:
```diff
@@ -276,21 +276,14 @@ ss::future<> sharded_store::delete_schema(schema_id id) {
       shard_for(id), _smp_opts, [id](store& s) { s.delete_schema(id); });
 }
 
-ss::future<stored_schema> sharded_store::has_schema(
-  subject_schema schema, include_deleted inc_del, normalize norm) {
+ss::future<stored_schema>
+sharded_store::has_schema(subject_schema schema, include_deleted inc_del) {
     auto versions = co_await get_versions(schema.sub(), inc_del);
 
-    try {
-        co_await validate_schema(schema.share());
-    } catch (const exception& e) {
-        throw as_exception(invalid_subject_schema(schema.sub()));
```
Do we get this nice error message if it fails validation at the callsite?

IoannisRP replied:
My argument was that we always call `make_canonical_schema` on inputs, thinking that this validation was part of it.

I was wrong cause not all 3 types of supported types have validation as part of canonicalization. So it is not redundant and it is needed.

Brought it back.

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/8ea23e00805ad82e22442c6c35a887e7375bc5ca..f434d7e6e81683227c037f30ff7cfe9275b45147):
- revert all changes to `transform-sdk`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64331-->
<!--c2facad8-d185-480e-b7f3-829e7e6721ee-->
<!--ab14d94e-6e86-4b6f-ab38-f1f73f7987f8-->
<!--c90ce6ba-f727-4c6d-9bfc-f0431ec622d4-->
<!--33972841-0d31-46cb-a704-5a602939ebce-->
<!--bf654955-3af9-4412-b1ab-5b170664b998-->
<!--e450a329-0191-4c80-801f-45b5552196e1-->
<!--title-->
## Retry command for [Build#64331](https://buildkite.com/redpanda/redpanda/builds/64331)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/schema_registry_test.py::SchemaRegistryAutoAuthTest.test_post_subjects_subject_versions
tests/rptest/tests/schema_registry_test.py::SchemaRegistryTest.test_always_normalize_option
tests/rptest/tests/schema_registry_test.py::SchemaRegistryTest.test_post_subjects_subject
tests/rptest/tests/schema_registry_test.py::SchemaRegistryAutoAuthTest.test_always_normalize_option
tests/rptest/tests/schema_registry_test.py::SchemaRegistryAutoAuthTest.test_post_subjects_subject
tests/rptest/tests/schema_registry_test.py::SchemaRegistryTest.test_post_subjects_subject_versions
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--afdb78c4-c46c-4e7e-bdef-d2d5678249b9-->
<!--TEST RESULTS-->
<!--831fb5c2-75c2-43c2-8198-09950b706ad0-->
<!--TEST RESULTS-->
<!--9c47d254-f95e-4715-b5c6-9cf9d621ed90-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64331">build#64331</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961be1-8212-4e1d-ac09-84f15c4c7633 | FLAKY | 1/2 |
| rptest.tests.idempotency_test.IdempotencySnapshotDelivery.test_recovery_after_snapshot_is_delivered | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961bdf-6fa0-4e81-8841-bd56e11abbe8 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961be1-8212-4e1d-ac09-84f15c4c7633 | FLAKY | 1/2 |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_always_normalize_option | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961bdf-6fa0-4e81-8841-bd56e11abbe8 | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_always_normalize_option | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961be1-8210-4fdc-ab49-15d8a5820b9e | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_post_subjects_subject | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961bdf-6fa0-4e81-8841-bd56e11abbe8 | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_post_subjects_subject | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961be1-8210-4fdc-ab49-15d8a5820b9e | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_post_subjects_subject_versions | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961bdf-6fa1-4177-8a52-d2028f86be4e | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_post_subjects_subject_versions | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961be1-8211-4f3f-96f3-27a0b20f69b3 | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryTest.test_always_normalize_option | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961bdf-6fa1-4177-8a52-d2028f86be4e | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryTest.test_always_normalize_option | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961be1-8211-4f3f-96f3-27a0b20f69b3 | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryTest.test_post_subjects_subject | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961bdf-6fa1-4177-8a52-d2028f86be4e | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryTest.test_post_subjects_subject | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961be1-8211-4f3f-96f3-27a0b20f69b3 | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryTest.test_post_subjects_subject_versions | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961bdf-6fa1-40a4-92dd-9a142c9fe122 | FAIL | 0/20 |
| rptest.tests.schema_registry_test.SchemaRegistryTest.test_post_subjects_subject_versions | ducktape | https://buildkite.com/redpanda/redpanda/builds/64331#01961be1-8212-4e1d-ac09-84f15c4c7633 | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64363">build#64363</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64363#01961fa7-c34e-4cdf-8bf9-841faf225cb3 | FLAKY | 1/2 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/64363#01961f95-5ab9-48c6-a729-ce15624cf6c2 | FLAKY | 1/3 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=True.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/64363#01961f95-5abc-4db6-a977-15ef5035fc87 | FLAKY | 1/2 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64363#01961f95-5ab9-48c6-a729-ce15624cf6c2 | FLAKY | 1/3 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64363#01961fa7-c34f-4cae-930c-71acf52e3ff0 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64369">build#64369</a></summary>


| test_id                                                                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| datalake_cloud_rpfixture.datalake_cloud_rpfixture                                                                                                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/64369#01962045-dc34-49fd-a2ad-622d19848b2d | FLAKY         | 1/2      |
| kafka_server_rpfixture.kafka_server_rpfixture                                                                                                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/64369#01962045-dc35-4b0a-bf95-553ee1c7d7f9 | FLAKY         | 1/2      |
| rptest.tests.archival_test.ArchivalTest.test_single_partition_leadership_transfer.cloud_storage_type=CloudStorageType.S3                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64369#0196208d-6115-487b-bd50-c4f0acf0034f | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64369#01962088-5625-47be-9997-5eac0bbcbc54 | FLAKY         | 1/2      |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64369#0196208d-6113-4f7a-91a0-999bfe970090 | FLAKY         | 1/2      |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64369#01962088-5626-4194-aabb-105942a75788 | FLAKY         | 1/5      |
</details>
<!--/body-->

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/f434d7e6e81683227c037f30ff7cfe9275b45147..b8b5531c3aa0d6d8919d5a9abfec896f53f1a20d):
- fix failing tests
- addressed review comments

BenPope commented on src/v/pandaproxy/schema_registry/storage.h:
```diff
@@ -505,8 +500,7 @@ class schema_value_handler final : public json::base_handler<Encoding> {
         case state::definition: {
-            _schema.def = typename typed_schema_definition<Tag>::raw_string{
-              ss::sstring{sv}};
+            _schema.def = schema_definition::raw_string{ss::sstring{sv}};
```
Drive-by fix, probably best to add it to another commit:
```suggestion
            _schema.def = schema_definition::raw_string{sv};

BenPope commented on src/v/pandaproxy/schema_registry/types.h:
```diff
@@ -230,15 +205,13 @@ class avro_schema_definition {
-    explicit operator canonical_schema_definition() const {
-        return {raw(), type()};
-    }
+    explicit operator schema_definition() const { return {raw(), type()}; }
```
Drive-by:
```suggestion
    explicit operator schema_definition() const {
        return {raw(), type(), refs()};
    }
```
This appears to only be used in tests.

BenPope commented on src/v/pandaproxy/schema_registry/sharded_store.cc:
```diff
@@ -298,6 +302,14 @@ ss::future<bool> sharded_store::upsert(
+    try {
+        subject_schema canonical = co_await make_canonical_schema(
+          schema.share(), normalize::no, false);
+        schema = std::move(canonical);
```
nitpick:
```suggestion
        schema = co_await make_canonical_schema(
          schema.share(), normalize::no, false);
```

BenPope reviewed (commented):
I'm seeing some leaks in `test_protobuf_rendering`

IoannisRP commented:
@BenPope what leaks are you referring to? None of the comments are for the `test/protobuf_rendering.cc` file :thinking:

BenPope commented:
> @BenPope what leaks are you referring to? None of the comments are for the `test/protobuf_rendering.cc` file 🤔

Turns out it was a bad build.

IoannisRP committed:
pandaproxy/sr: minor cleanup across sr

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25764-v24.3.x-746 remotes/upstream/v24.3.x
git cherry-pick -x 87d0471d54 e0a6210d4a f36aa4dfeb 4466736e1e b8b5531c3a 16c6d4ad3d
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14399276753).

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25764-v24.2.x-736 remotes/upstream/v24.2.x
git cherry-pick -x 87d0471d54 e0a6210d4a f36aa4dfeb 4466736e1e b8b5531c3a 16c6d4ad3d
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14399277282).
